zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+
+            let mut outcome_info =
+                <Outcomes<T>>::get(market_id, &outcome).ok_or(Error::<T>::OutcomeDoesNotExist)?;
+            let outcome_owners_len = outcome_info.owners.len() as u32;
+
+            // The `outcome_sum` never decreases (only increases) to allow
+            // caching the outcome with the highest `outcome_sum`.
+            // If the `outcome_sum` decreases, it would lead to more storage,
+            // because the winning outcome could have a smaller `outcome_sum`
+            // than the second highest `outcome_sum`.
+            let add_to_outcome_sum = |a| {
+                outcome_info.outcome_sum = outcome_info.outcome_sum.saturating_add(a);
+                <Winners<T>>::mutate(market_id, |highest| {
+                    *highest = Some(highest.clone().map_or(
+                        WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum),
+                        |prev_winner_info| {
+                            if outcome_info.outcome_sum >= prev_winner_info.vote_sum {
+                                WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum)
+                            } else {
+                                prev_winner_info
+                            }
+                        },
+                    ));
+                });
+                <Outcomes<T>>::insert(market_id, &outcome, outcome_info);
+            };
+
+            let vote_lock_counter = <Locks<T>>::try_mutate(
+                &sender,
+                |LockInfo(ref mut lock_info)| -> Result<u32, DispatchError> {
+                    match lock_info.binary_search_by_key(&market_id, |i| i.0) {
+                        Ok(i) => {
+                            let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
+                            // concious decision here to not throw an error when the new amount is not higher
+                            // because the user would have to pay worst case fees then
+                            if amount > prev_highest_amount {
+                                let diff = amount.saturating_sub(prev_highest_amount);
+                                add_to_outcome_sum(diff);
+                                lock_info[i].1 = amount;
+                            }
+                        }
+                        Err(i) => {
+                            ensure!(
+                                lock_info.len() as u32 <= T::MaxGlobalDisputeVotes::get(),
+                                Error::<T>::MaxVotesReached
+                            );
+                            add_to_outcome_sum(amount);
+                            lock_info.insert(i, (market_id, amount));
+                        }
+                    }
+                    Ok(lock_info.len() as u32)
+                },
+            )?;
+
+            T::Currency::extend_lock(
+                T::VoteLockIdentifier::get(),
```

Comment:
That sounds complicated to me 😅 How does it look in code?

Reply:
However the implementation might look, the idea was to assign one lock identifier per market. The proposed solution just concatenates 4 bytes of the market id to an unique identifier to avoid collisions. 
After thinking more about it, I don't think this proposed solution is suitable here, because it depends on `MaxLocks` of the balance pallet and consumes locks from the total pool of available locks. I think in this case, where we can potentially have huge amounts of vote locks, you implemented the best solution: Internal lock management.